util/txesbmantool/txesbmantool.c

```diff
@@ -0,0 +1,1407 @@
+/* SPDX-License-Identifier: GPL-2.0-only */
+
+#include <getopt.h>
+#include <stdbool.h>
+#include <stdio.h>
+#include <sys/stat.h>
+
+#ifndef WOLFSSL_USER_SETTINGS
+	#include <wolfssl/options.h>
+#endif
+
+#include <wolfssl/ssl.h>
+#include <wolfssl/wolfcrypt/rsa.h>
+#include <wolfssl/wolfcrypt/signature.h>
+#include <wolfssl/wolfcrypt/hash.h>
+#include <wolfssl/wolfcrypt/error-crypt.h>
+
+#ifndef ARRAY_SIZE
+#define ARRAY_SIZE(a) (sizeof(a) / sizeof((a)[0]))
+#endif
+
+#define WARN(...) fprintf(stderr, "WARNING: " __VA_ARGS__)
+#define ERROR(...) fprintf(stderr, "ERROR: " __VA_ARGS__)
+
+#define RSA_KEY_SIZE			2048
+#define RSA_KEY_EXPONENT		0x10001
+#define DER_FILE_BUFFER			2048 /* max DER size */
+
+#define MANIFEST_SIZE			1024
+
+#define SB_MANIFEST_MAGIC		0x4d425624	/* $VBM in LE */
+#define SB_MANIFEST_VERSION		0x1
+#define SB_MANIFEST_SIZE		MANIFEST_SIZE
+
+#define KEY_MANIFEST_MAGIC		0x4d4b5324	/* $SKM in LE */
+#define KEY_MANIFEST_VERSION		0x1
+#define KEY_MANIFEST_PAD_SIZE		(4096 - MANIFEST_SIZE)
+#define KEY_MANIFEST_RESERVED_SIZE	456
+#define KEY_MANIFEST_SIZE		(MANIFEST_SIZE + KEY_MANIFEST_PAD_SIZE)
+
+#define IBB_SIZE			(0x20000 - SB_MANIFEST_SIZE)
+
+struct manifest_header {
+	uint32_t magic;
+	uint32_t version;
+	uint32_t size;
+	uint32_t svn;
+} __attribute__((packed));
+
+struct key_signature {
+	uint8_t modulus[256];
+	uint32_t exponent;
+	uint8_t signature[256];
+} __attribute__((packed));
+
+struct sb_manifest {
+	struct manifest_header header;
+	uint32_t reserved1;
+	uint8_t ibb_hash[WC_SHA256_DIGEST_SIZE];
+	uint32_t reserved2;
+	uint32_t reserved3[8];
+	uint8_t oem_data[400];
+	uint8_t reserved4[20];
+	struct key_signature key_sig;
+} __attribute__((packed));
+
+struct key_manifest {
+	struct manifest_header header;
+	uint32_t km_id;
+	uint8_t sb_key_hash[WC_SHA256_DIGEST_SIZE];
+	uint8_t reserved[KEY_MANIFEST_RESERVED_SIZE];
+	struct key_signature key_sig;
+	uint8_t pad[KEY_MANIFEST_PAD_SIZE];
+} __attribute__((packed));
+
+typedef union {
+	struct sb_manifest sbm;
+	struct key_manifest km;
+} secure_boot_manifest;
```

Comment:
Is there a reason for making this union? AFAICT every time it is passed between functions, `enum manifest_type` is also passed, wouldn't it be easier to call separate functions for `sbm` and `km` instead?

Reply:
Dunno